Add Usage to docs (#76)
IAMconsortium/nomenclature: doc/source/usage.rst

.. _usage:

Usage
=====

The **DataStructureDefinition** and **RegionProcessor** classes
are initialized from yaml files that must follow specific formats.

> This page describes the required specifications of the files.

Author: Why should this be removed?

> This page describes the required specifications of the files.

Reviewer: In another comment I moved it up.

> This page describes the required specifications of the files.

Author: Ah ok, since the your previous comment referred to an outdated version of the code I didn't see it in the files tab.

      unit: A unit
      description: A short description
      <other attribute>: Some text (optional)

**Notes**

* Every variable **must have** a **unit**, which should be compatible with the
  Python package `iam-units <https://github.com/iamconsortium/units>`_.

* The unit attribute can also be empty, i.e., the variable is *dimensionless*.

* All other attributes such as "description" are optional. It is recommended to provide
  a description for better documentation.

* For region aggregation, the pyam function `aggregate_region
  <https://pyam-iamc.readthedocs.io/en/stable/api/iamdataframe.html#pyam.IamDataFrame.aggregate_region>`_
  is used. Attributes provided in a mapping that match the parameters "components",
  "method", "weight" or "drop_negative_weights" will be used in the aggregation.

Region
~~~~~~

To avoid repeating a “hierarchy” attribute many times (e.g., country,
continent), the yaml files must have a nested dictionary structure:

.. code:: yaml

   - <Hierarchy Level>:
     - region 1:
         Attribute: Attribute value
     - region 2:
         Attribute: Attribute value 

**Notes**

* Every region **must be** defined as part of a hierarchy. 

> * When importing the codelist, the hierarchy will be added as attribute,

Author: No this only applies to the *region* codelist.

  such that it can be retrieved as:

.. code:: python

   DataStructureDefinition.region["Region Name"]["Hierarchy"] = "<Hierarchy Level>"

* Other attributes specified in the yaml file can include (for countries)

>   ISO2/3-codes, or the list of countries included in a macro-region (i.e.,

Author: Ah yes I'll put a wikipedia link

  a continent or large region).

Tag
~~~

To avoid repetition (and subsequent errors), any number of yaml files
can be used as “tags” using a nested list of dictionaries.

There must be only one top-level entry in any yaml file to be used as
tag.

The files defining the tags must have a name starting with ``tag_``.

.. code:: yaml

   - <Tag>:
     - Some Key:
         description: a short description of the key

> When importing the codelist, any occurrence of ``<Tag>`` in a variable

Author: Good question, I don't think so, worded like this it makes it sound like there can only be one tag codelist which is not the case.
I'll change it to something like:
When importing a tag codelist, any occurrence of <Tag> in a variable

   well.
-  **model** (str): specifies the model name for which the mapping
   applies.
-  **native_regions** (list): list of model native regions serves as
   a selection as to which regions to keep.

   -  In the above example *region_a* is to be renamed to
      *alternative_name_a*. This is done by defining a key-value pair
      of *model_native_name: new_name*.
   -  *region_b* is selected but the name is not changed.
   -  Assuming *model_a* also defines a third region *region_c*,
      since it is not mentioned it will be **dropped** from the data.

-  **common_regions** (list): list of common regions which will be
   computed as aggregates. They are defined as list entries which
   themselves have a list of constituent regions. These constituent
   regions must be model native regions.

   -  **Important to note** the names of the constituent regions
      **must** refer to the **original** model native region names. In
      the above example *region_a* and *region_b* and **not**
      *alternative_name_a*.

Guidelines and variable naming conventions
------------------------------------------

The variable name (code) should adhere to the following conventions:

> -  A *|* (pipe) character indicates levels of hierarchy

Reviewer: ```suggestion
-  A *|* (pipe) character indicates levels of hierarchy.
```

-  Do not use spaces before and after the *|* character, but add a

>    space between words (e.g., *Primary Energy|Non-Biomass Renewables*)

Reviewer: ```suggestion
   space between words (e.g., *Primary Energy|Non-Biomass Renewables*).
```

> -  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.)

Reviewer: ```suggestion
-  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.).
```

> -  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary

Reviewer: ```suggestion
-  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary.
-  Do not use abbreviations of statistical operations (*min*, *max*, *avg*) - always spell out the word.
```

-  Add hierarchy levels where it might be useful in the future, e.g.,
   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
   Electric Vehicle*

> -  Do not use abbreviations of statistical operations (*min*, *max*,

Reviewer: ```suggestion
```

Author: Do you mean to delete this sentence or to move it up as well?

Reviewer: Yes, I'd move it up as well, like suggested in another suggestion - under the other comment on abbreviation maybe.

>    *avg*) but always spell out the word

Reviewer: ```suggestion
```

-  Do not include words like *Level* or *Quantity* in the variable,
   because this should be clear from the context or unit

> Class Overview

Reviewer: Fits best in API documentation, since "...(a full list of all classes can be found in API documentation)...."

Author: Removed this part as it was indeed not very useful here.

--------------

> A **DataStructureDefinition** class contains **CodeLists** for

Reviewer: **CodeLists** was written above small, we could either also use "codelists" here or above **CodeLists**. Or maybe I'm a bit confused ;)

Author: I removed this section as it was no longer necessary.

.. _usage:

Usage
=====

> Overview
> --------
> 
> This sections dives deeper into the inner workings of **DataStructureDefinition** and
> **RegionProcessor** as well as their associated auxiliary classes.
> 
> 
> Yaml file schemas
> --------------------
> 
> Both classes read information from yaml files which are required to follow specific
> structures.

Reviewer 2: Suggestion for a more concise, less verbose introduction.

```suggestion
The **DataStructureDefinition** and **RegionProcessor** classes
are initialized from yaml files that must follow specific formats.

This page describes the required specifications of the files.
```

> For **DataStructureDefinition** there are three types of schemas: variable, region and tag.

Reviewer 2: Suggest to add a TOC at the top of this document and use the class names as section heading

Author: Added a toc

  a description for better documentation.

* For region aggregation, the pyam function `aggregate_region
  <https://pyam-iamc.readthedocs.io/en/stable/api/iamdataframe.html#pyam.IamDataFrame.aggregate_region>`_
  is used. Attributes provided in a mapping that match the parameters "components",
  "method", "weight" or "drop_negative_weights" will be used in the aggregation.

Region
~~~~~~

To avoid repeating a “hierarchy” attribute many times (e.g., country,
continent), the yaml files must have a nested dictionary structure:

.. code:: yaml

   - <Hierarchy Level>:
     - region 1:
         Attribute: Attribute value
     - region 2:
         Attribute: Attribute value 

**Notes**

* Every region **must be** defined as part of a hierarchy. 

* When importing the codelist, the hierarchy will be added as attribute,
  such that it can be retrieved as:

.. code:: python

   DataStructureDefinition.region["Region Name"]["Hierarchy"] = "<Hierarchy Level>"

* Other attributes specified in the yaml file can include (for countries)
  ISO2/3-codes, or the list of countries included in a macro-region (i.e.,
  a continent or large region).

> 

Reviewer 2: "Generic" format should also be mentioned. This can be used, for example, to specify codelists for the allowed scenario names.

Author: Added generic to the list

Tag
~~~

To avoid repetition (and subsequent errors), any number of yaml files
can be used as “tags” using a nested list of dictionaries.

There must be only one top-level entry in any yaml file to be used as
tag.

The files defining the tags must have a name starting with ``tag_``.

.. code:: yaml

   - <Tag>:
     - Some Key:
         description: a short description of the key

When importing the codelist, any occurrence of ``<Tag>`` in a variable
name will be replaced by every element in the Tag dictionary. The
``<Tag>`` will also be replaced in any of the variable attributes.

> Mapping
> ~~~~~~~

Reviewer 2: ```suggestion
RegionProcessor
-----------------
```

Model mappings, defined on a per-model basis serve three different purposes:

1. Define a list of model native regions that are to be selected (and
   usually uploaded) from an IAM result. This also serves as an implicit
   exclusion list for model native regions, since only explicitly
   mentioned regions are selected.

2. Allow for renaming of model native regions.

3. Define how model native regions should be aggregated to common
   regions.

This example illustrates how such a model mapping looks like:

.. code:: yaml

   model: model_a
   native_regions:
     - region_a: alternative_name_a
     - region_b
   common_regions:
     - common_region_1:
       - region_a
       - region_b
     - common_region_2:
       - ...

**Notes**

-  The names of the three top level keywords **model**,
   **native_regions** and **common_regions** are fixed.

> -  Required properties are **model** and **at least** either

Reviewer 2: Use different formatting from the fixed keywords
```suggestion
-  Required properties are **model** and *at least* either
```

   **native_regions** or **common_regions**. **Both** are **allowed** as
   well.
-  **model** (str): specifies the model name for which the mapping
   applies.
-  **native_regions** (list): list of model native regions serves as
   a selection as to which regions to keep.

   -  In the above example *region_a* is to be renamed to
      *alternative_name_a*. This is done by defining a key-value pair
      of *model_native_name: new_name*.
   -  *region_b* is selected but the name is not changed.
   -  Assuming *model_a* also defines a third region *region_c*,
      since it is not mentioned it will be **dropped** from the data.

-  **common_regions** (list): list of common regions which will be
   computed as aggregates. They are defined as list entries which
   themselves have a list of constituent regions. These constituent
   regions must be model native regions.

   -  **Important to note** the names of the constituent regions
      **must** refer to the **original** model native region names. In
      the above example *region_a* and *region_b* and **not**
      *alternative_name_a*.

> Guidelines and variable naming conventions

Reviewer 2: Suggest to move this to an own page "Guidelines"

Author: Added a guidelines page

------------------------------------------

The variable name (code) should adhere to the following conventions:

-  A *|* (pipe) character indicates levels of hierarchy
-  Do not use spaces before and after the *|* character, but add a
   space between words (e.g., *Primary Energy|Non-Biomass Renewables*)
-  All words must be capitalised (except for *and*, *w/*, *w/o*, etc.)
-  Do not use abbreviations (e.g, *PHEV*) unless strictly necessary
-  Add hierarchy levels where it might be useful in the future, e.g.,
   use *Electric Vehicle|Plugin-Hybrid* instead of *Plugin-Hybrid
   Electric Vehicle*
-  Do not use abbreviations of statistical operations (*min*, *max*,
   *avg*) but always spell out the word
-  Do not include words like *Level* or *Quantity* in the variable,
   because this should be clear from the context or unit

> Class Overview

Reviewer 2: This section doesn't really make sense at the bottom. Best to integrate the descriptions into the text above.

Author: Will do, should I start using the linked references for every mention of a class like you started in #77?
So :class:`nomenclature.codelist.Codelist` instead of **CodeLists**?

Reviewer 2: Yes, we should use class-references wherever possible (that's why I started #77 today, to allow adopting this consistently right away).

Author: Done
